src/gui/layertree/qgslayertreeviewdefaultactions.cpp

```diff
@@ -429,27 +436,40 @@ void QgsLayerTreeViewDefaultActions::moveOutOfGroup()
 
 void QgsLayerTreeViewDefaultActions::moveToTop()
 {
-  QHash <QgsLayerTreeGroup *, int> groupInsertIdx;
-  int insertIdx;
-  const QList< QgsLayerTreeNode * >  selectedNodes = mView->selectedNodes();
-  for ( QgsLayerTreeNode *n : selectedNodes )
+  QList< QgsLayerTreeNode * >  selectedNodes = mView->selectedNodes();
+  // sort the nodes by depth first to avoid moving a group before it's contents
+  std::stable_sort( selectedNodes.begin(), selectedNodes.end(), []( const QgsLayerTreeNode * a, const QgsLayerTreeNode * b )
+  {
+    return a->depth() < b->depth();
+  } );
+  for ( auto it = selectedNodes.rbegin(); it < selectedNodes.rend(); ++it )
```

Comment:
Can we use an (easier to read) range-based for loop here?

Reply:
We need to iterate in reverse in order to preserve the selection order. If we use a range based for, we'll need to have the list reversed first.